### ISSUE TITLE
arrow-body-style: [2, "as-needed"] should ignore function bodies that return another block/function

### ISSUE THREAD

**Issue by bentruyman** [eslint/eslint#4353]
Granted the below example could be achieved a number of different ways to produce the same results, it's a use case that I think others might find useful if handled in the way I'm expecting.

The following code will fail **ESLint v1.9.0** with `arrow-body-style: [2, "as-needed"]`:

```javascript
function foo() {
  return (bar, baz) => { // Unexpected block statement surrounding arrow body. (arrow-body-style)
    return qux => {
      console.log(bar);
      console.log(baz);
      console.log(qux);
    };
  };
}

foo()("bar", "baz")("qux");
```

I expect the above (contrived currying example) to pass since, as far as I know, a block is required here.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Can you show the actual output from ESLint?

**Comment by michaelficarra** [comment]
The block is not required there.

**Comment by bentruyman** [comment]
Sorry, I should have posted more information in the original report.

Here's the a minimal `eslint` configuration to repro:

```yaml
---
env:
  es6: true
ecmaFeatures:
  arrowFunctions: true
rules:
  arrow-body-style: [2, "as-needed"]
```

Here's the source file:

```javascript
function foo() {
  return (bar, baz) => {
    return qux => {
      console.log(bar);
      console.log(baz);
      console.log(qux);
    };
  };
}

foo()("bar", "baz")("qux");
```

Here's the output from `eslint`:

```sh
/Users/someone/Desktop/eslint/index.js
  2:24  error  Unexpected block statement surrounding arrow body  arrow-body-style

✖ 1 problem (1 error, 0 warnings)
```

Is there a way to write this to not generate the failure?

**Comment by michaelficarra** [comment]
```js
function foo() {
  return (bar, baz) =>
    qux => {
      console.log(bar);
      console.log(baz);
      console.log(qux);
    };
}

foo()("bar", "baz")("qux");
```

**Comment by bentruyman** [comment]
I can confirm what @michaelficarra posted does remove the error. While it's awkward to write code like this, in my opinion, it's probably unlikely another would run into this.

Unless this use case seems useful (I doubt it is) this should probably be closed.

I imagine most developers would break out something like this into individual functions in which case this wouldn't be an issue.

```javascript
function foo() {
  return (bar, baz) => thenQux.bind(null, bar, baz);
}

function thenQux(bar, baz, qux) {
  console.log(bar);
  console.log(baz);
  console.log(qux);
}

foo()("bar", "baz")("qux");
```

**Comment by ilyavolodin** [comment]
I agree. I think in most cases for readability you would want separate functions. Closing as there is a few workarounds available.

**Comment by mattkrick** [comment]
I think this is worth a 2nd look, since it's a pattern that comes up often when using higher order functions, like HOCs and middlewares. Personally, removing the blocks or breaking apart a HOC is more confusing than just writing something like:
```
const hoc = options => ComposedComponent => {
  return class ExtendedComponent extends Component {...
```

**Comment by Bartekus** [comment]
I'll second @mattkrick on that, and provide practical example in form of Login.js component:
```
import React, { PropTypes } from 'react';

const Login = (props) => {
  return <a onClick={props.onClick}>Login</a>;
};

Login.propTypes = {
  onClick: PropTypes.func.isRequired,
};

export default Login;
```
There is definitely need for such pattern and that need will only increase as we further push the envelope of possibilities that react&redux for example allow.

Please reconsider you position on this.

**Comment by elgubenis** [comment]
Hey, I'm not sure what I am doing wrong, but somehow found this Issue related, I get an eslint error in these examples:

```js
    const commentNodes = this.props.data.map((comment) => {
      return <Comment author={comment.author} key={comment.id}>{comment.text}</Comment>;
    });
```

```js
    const commentNodes = this.props.data.map((comment) => {
      return (
        <Comment author={comment.author} key={comment.id}>
          {comment.text}
        </Comment>
      );
    });
```

could someone be so kind to give me a little hint? the error is: ```Unexpected block statement surrounding arrow body.```

**Comment by mysticatea** [comment]
@elgubenis It's [arrow-body-style](http://eslint.org/docs/rules/arrow-body-style)'s warning.
The arrow function of your code has only a `return` statement in the body, so you can rewrite them with `() => foo` style.

**Comment by tribou** [comment]
Encountered this issue writing a GraphQL schema objects with resolve functions that return promises.  I think setting `arrow-body-style` to `always` makes the code more readable through consistency.  Having arrow functions take on different forms throughout similar blocks of code just adds more context overhead.  Am I missing the benefit?
